lib/Conversion/TorchToLinalg/Pooling.cpp

```diff
@@ -228,6 +228,90 @@ class ConvertAtenAdaptiveAvgPool2dOp
+    : public OpConversionPattern<AtenMaxPool2dWithIndicesBackwardOp> {
+public:
+  using OpConversionPattern::OpConversionPattern;
+  LogicalResult
+  matchAndRewrite(AtenMaxPool2dWithIndicesBackwardOp op, OpAdaptor adaptor,
+                  ConversionPatternRewriter &rewriter) const override {
+    if (failed(verifyLinalgCompatibleTypes(op, rewriter)))
+      return failure();
+    Location loc = op->getLoc();
+    Value gradOutput = adaptor.grad_output();
+    Value self = adaptor.self();
+    Value indices = adaptor.indices();
+    RankedTensorType gradOutputType =
+        gradOutput.getType().cast<RankedTensorType>();
+    unsigned gradOutputRank = gradOutputType.getRank();
+    RankedTensorType newResultRankedTensorType =
+        getTypeConverter()->convertType(op.getType()).cast<RankedTensorType>();
+
+    // Creating a new tensor with the size same as self but last two dimensions
+    // collapsed into a single dimension.
+    SmallVector<Value> resultShape = getTensorSizes(rewriter, loc, self);
+    Value wIn = resultShape[gradOutputRank - 1];
+
+    Value result = createZeroInitTensor(
+        rewriter, loc, resultShape, newResultRankedTensorType.getElementType());
+
+    AffineExpr dimExpr = rewriter.getAffineDimExpr(0);
+    SmallVector<AffineMap, 2> indexingMaps(
+        2, rewriter.getMultiDimIdentityMap(gradOutputRank));
+    indexingMaps.push_back(
+        AffineMap::get(gradOutputRank, 0, dimExpr, op->getContext()));
+    SmallVector<StringRef> iteratorTypes(gradOutputRank - 2,
+                                         getParallelIteratorTypeName());
+    iteratorTypes.push_back(getReductionIteratorTypeName());
+    iteratorTypes.push_back(getReductionIteratorTypeName());
+
+    // Creating a dummy tensor to be iterated over by generic op for the op
+    // requirements.
+    Value dummyTensor = rewriter.create<linalg::InitTensorOp>(
+        loc, resultShape[0], newResultRankedTensorType.getElementType());
+
+    rewriter.create<linalg::GenericOp>(
+        loc, dummyTensor.getType(), ValueRange{gradOutput, indices},
+        ValueRange{dummyTensor}, indexingMaps, iteratorTypes,
+        [&](OpBuilder &b, Location loc, ValueRange args) {
+          Value out = args[0], index = castIntToIndex(b, loc, args[1]);
+          SmallVector<Value> resultIndices;
+          for (unsigned i = 0; i < gradOutputRank - 2; i++)
+            resultIndices.push_back(b.create<linalg::IndexOp>(loc, i));
+          resultIndices.push_back(
+              b.create<arith::FloorDivSIOp>(loc, index, wIn));
+          resultIndices.push_back(b.create<arith::RemUIOp>(loc, index, wIn));
+          Value temp = b.create<tensor::ExtractOp>(loc, result, resultIndices);
+          if (gradOutputType.getElementType().isa<mlir::FloatType>())
+            out = b.create<arith::AddFOp>(loc, temp, out);
+          else
+            out = b.create<arith::AddIOp>(loc, temp, out);
+
+          b.create<tensor::InsertOp>(loc, out, result, resultIndices);
```

Comment:
`tensor::InsertOp` can't be used here because this wont update the result tensor inplace but return a new tensor. See https://mlir.llvm.org/docs/Dialects/TensorOps/#tensorinsert-mlirtensorinsertop. `It returns a copy of dest with the proper slice updated with the value of scalar.` I think Scatter op should be used instead.

Reply:
I have made the changes to the code accordingly to use the scatter op.